diff --ruN a/stablehlo/stablehlo/dialect/StablehloOps.cpp b/stablehlo/stablehlo/dialect/StablehloOps.cpp
--- stablehlo/stablehlo/dialect/StablehloOps.cpp
+++ stablehlo/stablehlo/dialect/StablehloOps.cpp
@@ -129,6 +129,15 @@
       getBounds(), RankedTensorType::get(shape, elementType), emitError);
 }
 
+ArrayAttr PrecisionConfigAttr::get(MLIRContext* context,
+                                   ArrayRef<Precision> precisions) {
+  SmallVector<Attribute> precisionAttrs =
+      llvm::map_to_vector(precisions, [&](Precision precision) -> Attribute {
+        return PrecisionAttr::get(context, precision);
+      });
+  return ArrayAttr::get(context, precisionAttrs);
+}
+
 //===----------------------------------------------------------------------===//
 // ReduceScatterOp
 //===----------------------------------------------------------------------===//
diff --ruN a/stablehlo/stablehlo/dialect/StablehloOps.h b/stablehlo/stablehlo/dialect/StablehloOps.h
--- stablehlo/stablehlo/dialect/StablehloOps.h
+++ stablehlo/stablehlo/dialect/StablehloOps.h
@@ -40,6 +40,7 @@
 #include "mlir/IR/Types.h"
 #include "mlir/Interfaces/InferTypeOpInterface.h"
 #include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "mlir/Support/LLVM.h"
 #include "mlir/Support/LogicalResult.h"
 #include "stablehlo/dialect/Base.h"
 #include "stablehlo/dialect/Version.h"
@@ -191,6 +192,13 @@
   builder.create<stablehlo::ReturnOp>(loc, reducer.getResult());
 }
 
+// PrecisionConfigAttr is a constraint attribute on ArrayAttrs.
+// Create this class to allow for building this attr similar to other
+// attributes.
+struct PrecisionConfigAttr : public ArrayAttr {
+  static ArrayAttr get(MLIRContext *context, ArrayRef<Precision> precisions);
+};
+
 }  // end namespace stablehlo
 }  // end namespace mlir
 

